### PR TITLE
Update template with changes from sunpy core

### PR DESCRIPTION
**Pull request by Cadair**
This is a remake of #6 after I borked it. Hard.

TODO:
- [x] Merge in upstream
- [x] Set travis to default to no.
- [x] Update CirlceCI config with changes.

### CODE REVIEW

File: {{ cookiecutter.package_name }}/.circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+  name: Install Tox
+  command: |
+    pip install tox
+
+tox-conda-install: &tox-conda-install
+  name: Install Tox-Conda
+  command: |
+    pip install tox-conda
+
+image-run: &image-tests
```
Comment (Cadair) [inline {{ cookiecutter.package_name }}/.circleci/config.yml]: we probably don't need this in the template?

File: {{ cookiecutter.package_name }}/.circleci/config.yml [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+  environment:
+    MPLBACKEND: Agg
+  command: |
+    tox -e figure
+    pip install codecov
+    codecov
+
+version: 2
+jobs:
+  egg-info-36:
```
Comment (Cadair) [inline {{ cookiecutter.package_name }}/.circleci/config.yml]: we should drop this job either completely or conditionally if the minimum python version is 3.7

File: {{ cookiecutter.package_name }}/.circleci/config.yml [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+      - image: circleci/python:3.7
+    steps:
+      - checkout
+      - run: *skip-check
+      - run: *merge-check
+      - run: python setup.py egg_info
+
+  pip-install:
+    docker:
+      - image: continuumio/miniconda3
```
Comment (Cadair) [inline {{ cookiecutter.package_name }}/.circleci/config.yml]: (a bit off topic) but why are we using miniconda for the pip test?

File: {{ cookiecutter.package_name }}/.circleci/config.yml [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+      - run:
+          name: Install {{ cookiecutter.module_name }}
+          command: |
+            pip install --progress-bar off .[all,dev]
+            pip install --no-use-pep517 -e .[all,dev]
+            python setup.py develop
+
+  html-docs:
+    docker:
+      - image: continuumio/miniconda3
```
Comment (Cadair) [inline {{ cookiecutter.package_name }}/.circleci/config.yml]: same here actually this is just tox right?

File: {{ cookiecutter.package_name }}/.circleci/config.yml [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+      - run:
+          name: "Built documentation is available at:"
+          command: DOCS_URL="${CIRCLE_BUILD_URL}/artifacts/${CIRCLE_NODE_INDEX}/${CIRCLE_WORKING_DIRECTORY/#\~/$HOME}/docs/_build/html/index.html"; echo $DOCS_URL
+
+workflows:
+  version: 2
+
+  egg-info:
+    jobs:
+      - egg-info-36
```
Comment (Cadair) [inline {{ cookiecutter.package_name }}/.circleci/config.yml]: same conditional here.

File: {{ cookiecutter.package_name }}/pyproject.toml [hunk 6]
```diff
@@ -0,0 +1,88 @@
+[build-system]
+requires = ["setuptools", "setuptools_scm", "wheel", "sunpy"]
```
Comment (Cadair) [inline {{ cookiecutter.package_name }}/pyproject.toml]: I don't think we need `sunpy` here I think we need the hard pinned numpy. Unless there is a reason the package template needs to be able to import sunpy during `setup.py` which it really shouldn't

File: {{ cookiecutter.package_name }}/setup.cfg [hunk 7]
```diff
@@ -1,17 +1,52 @@
...
+license = {{ cookiecutter.license }}
+url = {{ cookiecutter.project_url }}
+edit_on_github = {{ cookiecutter.edit_on_github_extension }}
+github_project = {{ cookiecutter.github_project }}
+
+[options]
+python_requires = >={{ cookiecutter.minimum_python_version }}
+packages = find:
+include_package_data = True
+setup_requires = setuptools_scm
```
Comment (Cadair) [inline {{ cookiecutter.package_name }}/setup.cfg]: Are we moving the template to `setuptools_scm` with this? if so there are a bunch more changes we should make, like more docs, dropping `astropy_helpers` for python only projects etc. I would suggest we do that in a separate PR?

File: hooks/post_gen_project.py [hunk 8]
```diff
@@ -74,7 +75,7 @@ def process_licence(licence_name):
                 Repo.create_submodule(
                     new_repo, "astropy_helpers", "astropy_helpers",
                     "https://github.com/astropy/astropy-helpers.git",
-                    "{{ cookiecutter.astropy_helpers_version }}")
+                    "master")
```
Comment (nabobalis) [inline hooks/post_gen_project.py]: This change probably needs to be undone.

File: {{ cookiecutter.package_name }}/.circleci/config.yml (hunk 1 above)
Comment (nabobalis) [reply {{ cookiecutter.package_name }}/.circleci/config.yml]: I thought I removed all this. Woops.

File: {{ cookiecutter.package_name }}/.circleci/config.yml (hunk 2 above)
Comment (nabobalis) [reply {{ cookiecutter.package_name }}/.circleci/config.yml]: The min is 3.6?

File: {{ cookiecutter.package_name }}/setup.cfg (hunk 7 above)
Comment (nabobalis) [reply {{ cookiecutter.package_name }}/setup.cfg]: Well I just wanted to get the version working with setuptools.

File: {{ cookiecutter.package_name }}/.circleci/config.yml (hunk 3 above)
Comment (nabobalis) [reply {{ cookiecutter.package_name }}/.circleci/config.yml]: It worked?

File: {{ cookiecutter.package_name }}/.circleci/config.yml (hunk 4 above)
Comment (nabobalis) [reply {{ cookiecutter.package_name }}/.circleci/config.yml]: It worked so why change it?

Comment (Cadair) [reply {{ cookiecutter.package_name }}/.circleci/config.yml]: the miniconda image is much larger and not cached by default on the circleci workers so it will make the build a bit slower. Not that it matters that much but just a thought.

File: {{ cookiecutter.package_name }}/.circleci/config.yml (hunk 2 above)
Comment (Cadair) [reply {{ cookiecutter.package_name }}/.circleci/config.yml]: the minimum can be set by the user during the cookiecutter step. It's a jinja template variable.

**Comment by nabobalis** [comment]
I think this is ok to go. I need to actually test it when I create a package and setup the CI for it to see if it works.

File: {{ cookiecutter.package_name }}/tox.ini [hunk 9]
```diff
@@ -0,0 +1,26 @@
+[tox]
+envlist = py{36,37}-{offline,online,astropydev,numpydev,sunpydev},build_docs
+isolated_build = False
+
+[testenv]
+changedir = tmp
+setenv =
+    MPLBACKEND = agg
+    COLUMNS = 180
+    PYTEST_COMMAND = pytest --pyargs {{ cookiecutter.module_name }} --cov={{ cookiecutter.module_name }} --cov-config={toxinidir}/setup.cfg --verbose -m "not figure" --durations=25
```
Comment (Cadair) [inline {{ cookiecutter.package_name }}/tox.ini]: We should only use pyargs here if the package is templated with C extensions enabled. x/ref: https://github.com/sunpy/ndcube/pull/172/files#r287349133

Comment (nabobalis) [reply {{ cookiecutter.package_name }}/tox.ini]: Updated it.

**Comment by nabobalis** [comment]
it doesn't setup an example gallery, do we want to? add an option?

**Comment by nabobalis** [comment]
Merged in Upstream.

**Comment by nabobalis** [comment]
Well that was a bit more work and I had to do some strange things with __all__ to get the docs building.

File: {{ cookiecutter.package_name }}/azure-pipelines.yml [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+    tox: py36-offline --
+{%- endif %}
+
+- template: azure-templates.yml
+  parameters:
+    name: macOS_37_offline
+    os: macos
+    tox: py37-offline --
+
+- template: azure-templates.yml
```
Comment (Cadair) [inline {{ cookiecutter.package_name }}/azure-pipelines.yml]: This job uses 3.6 do we want it to be 3.7 if that's the min version.

Comment (nabobalis) [reply {{ cookiecutter.package_name }}/azure-pipelines.yml]: We can't template azure files due to the sythanx being the same as cookie cutter.

Comment (Cadair) [reply {{ cookiecutter.package_name }}/azure-pipelines.yml]: ah.
